Milestone 1: Default Dashboard Body to a Fixed Width

JesseSDevaney opened the pull request:
### Problem

Responsively sized dashboards lead to consumers seeing a dashboard in a way the creator did not intend for it to be consumed. The creator may have built their dashboard for consumption on a large screen only to have to their consumers view it on a small screen, potentially losing a lot of the useful value the creator intended the dashboard to have. 

### Value
This PR introduces a new `width` setting for dashboards. It has two values:
 1. `"full" = responsive`: the current behavior
2. `"fixed" = 1048px`: the default behavior for new dashboards

Creators will now be able to toggle this setting, build their dashboard, and have their consumers see it exactly the way they intended it to be seen.


---
### Desired Behavior/Acceptance Criteria
**Solution Behaviors**
- (in edit mode) Expose new ellipsis button that opens a popover filled with new settings options (at this time, only 1)
  ![image](https://github.com/metabase/metabase/assets/22608765/748561c8-d21f-42f1-af6a-afeca2d66c6e)
  ![image](https://github.com/metabase/metabase/assets/22608765/7b273e74-a1c5-4376-9ca9-d31b654cead1)
  - [ ] Expose width setting toggle in popover for changing dashboard width between "full" and "fixed".
    - [ ] Toggled on = "full"
    - [ ] Toggled off = "fixed"
- Dashboard body:
  - [ ] When width = "full", dashboard grid should auto-scale to the full-width available
    ![image](https://github.com/metabase/metabase/assets/22608765/35e8ac7c-3167-4bbd-a40b-e5aee487efb7)
  - [ ] When width = "fixed", dashboard grid should remain at the fixed width (1048px) until it is the size of the viewport and then should responsively downscale until reaching the mobile view
    ![image](https://github.com/metabase/metabase/assets/22608765/3d388af1-09bf-406d-ad25-d77ee536ea24)
    ![image](https://github.com/metabase/metabase/assets/22608765/41fd37fa-8d6f-45cb-abaa-a38d09e3b2c2)
    ![image](https://github.com/metabase/metabase/assets/22608765/7636fb8e-e3a1-43d6-bdec-5b3287dbdd5c)

  - [ ] Both options should condense to one column when the normal mobile breakpoint is reached
- [ ] Newly created dashboards should default to `width = "fixed"` 
- [ ] Previously existing dashboards should retain `width = "full"`

**Validation Tests**
- [ ] (`dashboard.cy.spec.js`) Validate new width setting works as intended
- [ ] (`public-dashboard.cy.spec.js`) Validate public dashboards respect the new "width" setting
- [ ] (`x-rays.cy.spec.js`) Validate X-ray dashboards are created with and respect `width = "fixed"`

---
### How to Manually Verify the Solution
**Full-App**
1. Checkout branch and start development server
2. Open an existing dashboard and see that it still has `width = "full"`
3. Create a new dashboard and see that it defaults to `width = "fixed"`
4. Toggle the `width` setting, save it, and refresh. See that the `width` setting persists after saving and refreshing.

**X-Rays**
1. In the collection sidebar, go to Browse Data > Databases > Sample Database > X-Ray on Table
  ![image](https://github.com/metabase/metabase/assets/22608765/dafaf068-d883-47a0-a757-5c2e0f40b642)
2. See that X-Ray dashboard body has fixed-width.
  ![image](https://github.com/metabase/metabase/assets/22608765/c96c8f98-d624-4864-9aee-e811e3fac155)

**Embedding**
1. Open up a dashboard in the full-app and enable embedding
  ![image](https://github.com/metabase/metabase/assets/22608765/246f87cb-11a9-40cf-b3e7-ea754fe77b15)
  ![image](https://github.com/metabase/metabase/assets/22608765/5297fd3a-bb08-4779-956d-7d7373514965)
4. Follow this guide to setup the rest of the static embedding app
   - https://github.com/metabase/metabase-nodejs-static-embedding-sample
5. Change the width toggle in the full-app, refresh the static embedding app, and see that the dashboard width is `"fixed"` when set to `"fixed"` and `"full"` when set to `"full"`

---
**Context**
- Completes Milestone 1: https://github.com/metabase/metabase/issues/36358

adam-james-v committed:
Dashboard PUT api endpoint accepts width changes and updates appdb

update-dashboard function now is aware of the :width key so those changes can end up in the transaction.

Also added a width test that asserts that the value's default is "fixed", it can be changed, eg. to "full", but cannot
be changed to other values.

adam-james-v committed:
Migration adding 'width' to Dashboards

3 migrations:
 - 1st adding the width column with default value of 'fixed'
 - 2nd updating all existing dashboards to have width 'full', which corresponds to what the current behaviour is (will
 be the 'old' behaviour after the fixed-width project lands).
   - The rolloback here is necessary but we don't care what happens as the column will be dropped immediately in the
   next rollback anyway
 - 3rd sets the notNullableConstraint. DefaultNull is 'full' here, just in case there's an existing dashboard whose
 width value is not yet set from the 1st migration. Don't know how that could happen, but its here in case

adam-james-v committed:
Add width to revision tests

adam-james-v committed:
Fix dashboard revision tests.

:width key is now needed in some revision tests. As well we need a string communicating that the :width setting has
changed from 'full' to 'fixed' or vice-versa.

adam-james-v committed:
Fix comments/remarks in migration to be accurate

adam-james-v committed:
Merge branch 'master' into fixed-width-dashboard-migration

adam-james-v committed:
Attempt to fix default not working mysql/mariadb

adam-james-v committed:
Merge branch 'fixed-width-dashboard-migration' of https://github.com/metabase/metabase into fixed-width-dashboard-migration

adam-james-v committed:
Set default in dashboard model

Signed-off-by: Adam James <adam.vermeer2@gmail.com>

adam-james-v committed:
Revert default :width 'fixed' value.

JesseSDevaney committed:
dashboard fixed width FE implementation

JesseSDevaney committed:
adjust popover shadow styling as the popover was blending in with the header

adam-james-v committed:
Explicitly add default value 'fixed' for MySQL/MariaDB

JesseSDevaney committed:
adjust extra button popover offset

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

JesseSDevaney committed:
add E2E to validate behavior

JesseSDevaney committed:
add fixed width container to public dashboards

JesseSDevaney committed:
fix public embedding not respecting dashboard width setting

JesseSDevaney committed:
add test for public dashboards

JesseSDevaney commented on frontend/src/metabase-types/api/dashboard.ts:
```diff
@@ -38,6 +38,7 @@ export interface Dashboard {
   archived: boolean;
   public_uuid: string | null;
   embedding_params?: Record<string, string> | null;
+  width: "full" | "fixed";
```
Add new dashboard attribute to dashboard type

JesseSDevaney commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.tsx:
```diff
@@ -138,6 +138,18 @@ export const ParametersWidgetContainer = styled(FullWidthContainer)<{
     `}
 `;
 
+export const FIXED_WIDTH = "1048px";
+export const FixedWidthContainer = styled.div<{
+  isFixedWidth: boolean;
+}>`
+  margin: 0 auto;
+  ${({ isFixedWidth }) =>
+    isFixedWidth &&
+    css`
+      max-width: ${FIXED_WIDTH};
+    `}
+`;
+
```
Create the fixed width container and use `isFixedWidth` variable to dynamically set the width

JesseSDevaney commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:
```diff
@@ -523,10 +524,14 @@ function DashboardInner(props: DashboardProps) {
-              <CardsContainer id="Dashboard-Cards-Container">
-                {renderContent()}
-              </CardsContainer>
+              <FixedWidthContainer
+                data-testid="fixed-width-container"
+                isFixedWidth={dashboard?.width === "fixed"}
+              >
+                <CardsContainer id="Dashboard-Cards-Container">
+                  {renderContent()}
+                </CardsContainer>
+              </FixedWidthContainer>
```
Wrap `FixedWidthContainer` around the dashboard cards body. Use `dashboard.width` attribute to dynamically set a fixed/responsive width.

JesseSDevaney commented on frontend/src/metabase/dashboard/components/DashboardButtonTooltip.tsx:
```diff
@@ -0,0 +1,30 @@
+import type { ReactNode } from "react";
+import { Text, Tooltip } from "metabase/ui";
+
+type DashboardButtonTooltipProps = {
+  children: ReactNode;
+  label: string;
+};
+
+export function DashboardButtonTooltip({
+  children,
+  label,
+}: DashboardButtonTooltipProps) {
+  return (
+    <Tooltip
+      py="0.6rem"
+      px="0.75rem"
+      bg="bg-black"
+      offset={4}
+      label={
+        <Text c="inherit" size="sm" fw={700}>
+          {label}
+        </Text>
+      }
+      withArrow
+      arrowSize={10}
+    >
+      {children}
+    </Tooltip>
+  );
+}
```
Update dashboard header button tooltips to use the new Mantine `Tooltip` component

kulyk replied:
Nice! 😍 QQ: did you have to apply all of these styles to make the tooltips look exactly as before?

JesseSDevaney replied:
I actually got this from somewhere else where someone converted the old tool-tip to use the new Mantine tooltip (https://github.com/metabase/metabase/blob/master/frontend/src/metabase/dashboard/components/DashboardEmbedHeaderButton/DashboardEmbedHeaderButton.tsx#L39-L51)

JesseSDevaney commented on frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx:
```diff
@@ -425,18 +432,30 @@ class DashboardHeaderContainer extends Component<DashboardHeaderProps> {
         );
       }
 
+      // Extra Buttons Menu
+      buttons.push(
+        <ExtraEditButtonsMenu
+          key="extra-options-button"
+          setDashboardAttribute={setDashboardAttribute}
+          dashboard={dashboard}
+        />,
+      );
+
```
Add the new edit mode button

JesseSDevaney commented on frontend/src/metabase/public/containers/PublicDashboard.jsx:
```diff
@@ -166,14 +167,19 @@ class PublicDashboard extends Component {
-                  metadata={this.props.metadata}
-                  navigateToNewCardFromDashboard={() => {}}
-                />
+                <FixedWidthContainer
+                  data-testid="fixed-width-container"
+                  isFixedWidth={dashboard?.width === "fixed"}
+                >
+                  <DashboardGridConnected
+                    {...this.props}
+                    isPublic
+                    className="spread"
+                    mode={PublicMode}
+                    metadata={this.props.metadata}
+                    navigateToNewCardFromDashboard={() => {}}
+                  />
+                </FixedWidthContainer>
```
Setup `FixedWidthContainer` on public dashboards

JesseSDevaney commented on frontend/src/metabase/ui/theme.ts:
```diff
@@ -38,7 +38,7 @@ export const getThemeOverrides = (): MantineThemeOverride => ({
   primaryShade: 0,
   shadows: {
-    sm: "0px 4px 4px 0px rgba(0, 0, 0, 0.08)",
+    sm: "0px 1px 4px 2px rgba(0, 0, 0, 0.08)",
```
Adjust shadow styling to match the designs styling. The other shadow styling did not create a shadow on the top of the popover making it blend in with the dashboard header

JesseSDevaney commented on src/metabase/api/public.clj:
```diff
@@ -231,7 +231,7 @@
   {:pre [(even? (count conditions))]}
   (binding [params/*ignore-current-user-perms-and-return-all-field-values* true]
-    (-> (api/check-404 (apply t2/select-one [Dashboard :name :description :id :parameters :auto_apply_filters], :archived false, conditions))
+    (-> (api/check-404 (apply t2/select-one [Dashboard :name :description :id :parameters :auto_apply_filters :width], :archived false, conditions))
```
Add new dashboard `width` attribute to public dashboard API response

adam-james-v committed:
Fix embed test failures

adam-james-v commented on test/metabase/api/embed_test.clj:
```diff
@@ -155,7 +155,7 @@
 (def successful-dashboard-info
   {:auto_apply_filters true, :description nil, :parameters [], :dashcards [], :tabs [],
-   :param_values {}, :param_fields nil})
+   :param_values {}, :param_fields nil :width "fixed"})
```
Adding the :width "fixed" default here should allow the tests to pass. Since we now send that key via the api, it's fair to add it as part of the expected successful dashboard info.

kulyk commented on e2e/test/scenarios/sharing/public-dashboard.cy.spec.js:
```diff
@@ -181,4 +181,32 @@ describe("scenarios > public > dashboard", () => {
+    );
+
+    // toggle full-width
+    cy.get("@dashboardId").then(id => {
+      cy.signInAsAdmin();
+      cy.request("PUT", `/api/dashboard/${id}`, {
+        width: "full",
+      });
+      visitPublicDashboard(id);
+    });
```
Do we actully need `cy.signInAsAdmin()` here?

JesseSDevaney replied:
It was failing the request to change the width, so I had to add that in there

kulyk commented on frontend/src/metabase/public/containers/PublicDashboard.jsx:
```diff
@@ -166,14 +167,19 @@ class PublicDashboard extends Component {
-                  metadata={this.props.metadata}
-                  navigateToNewCardFromDashboard={() => {}}
-                />
+                <FixedWidthContainer
+                  data-testid="fixed-width-container"
+                  isFixedWidth={dashboard?.width === "fixed"}
+                >
+                  <DashboardGridConnected
+                    {...this.props}
+                    isPublic
+                    className="spread"
+                    mode={PublicMode}
+                    metadata={this.props.metadata}
+                    navigateToNewCardFromDashboard={() => {}}
+                  />
```
@JesseSDevaney WDYT about making `FixedWidthContainer` a part of `DashboardGridConnected`?

We have three top-level dashboard components (`DashboardApp,` `PublicDashboard`, and `AutomaticDashboardApp`), and they all use `DashboardGridConnected`. I think it'd make sense to move the width logic there, so all three dashboard components can get this featyre for free

JesseSDevaney replied:
Great call-out. I moved the `FixedWidthContainer` into `DashboardGridConnected`

kulyk commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.tsx:
```diff
@@ -138,6 +138,18 @@ export const ParametersWidgetContainer = styled(FullWidthContainer)<{
     `}
 `;
 
+export const FIXED_WIDTH = "1048px";
```
Not sure if it's expected or not, but even though we're setting 1048px, the actual content is a bit narrower because of some horizontal margin. I think we'd want to remove it @kdoh @cdeweyx ? I want these 28px back! 😂 

![CleanShot 2024-02-02 at 18 09 50@2x](https://github.com/metabase/metabase/assets/17258145/a80e3f61-3d3b-4847-9d18-154eb138b1c5)

JesseSDevaney replied:
After I moved `FixedWidthContainer` into `DashboardGridConnected`, this is no longer an issue. It still has padding when it reaches the edges of the viewport, but that is where we want it.

kulyk reviewed (commented):
Love this 😍 Left couple of comments

JesseSDevaney committed:
add fixed width to x-ray dashboards

JesseSDevaney committed:
reduce code duplication

JesseSDevaney committed:
move FixedWidthContainer into DashboardGridConnected

- We get the fixed-width w/o code duplication across AutomaticDashboardApp, PublicDashboard, and Dashboard

JesseSDevaney committed:
update E2E tests to reflect actual intended behavior

JesseSDevaney committed:
fix type errors

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

adam-james-v commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -702,7 +702,8 @@
           :param_fields (filter-referenced-fields root query-filter)
-          :auto_apply_filters true))))
+          :auto_apply_filters true
+          :width "fixed"))))
```
Looked into it, and that’s as good a place as any to add the key.
One difference here, just to keep a good mental model, is that automagic dashboards are ephemeral. We are not pulling anything from the AppDB, and (as far as I understand it), once you hit ‘save’, it’s just a normal dashboard. So, the key you added is part of generating this ‘ephemeral’ dashboard, not requesting an additional column from the db.

The embed change we made pulls already existing data from the AppDB.

This doesn’t affect your change in any way, I just wanted to explain the difference between the embedded dashboard API and the automagic Dashboards.
I’ll add this as a review on the PR too. :slightly_smiling_face:

kulyk reviewed (approved):
Code looks good! 🚢 🇮🇹

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

JesseSDevaney committed:
adjust tooltip button after merging changes from master

JesseSDevaney committed:
fix bar chart test failure

JesseSDevaney committed:
fix dashboard filters date test failure

- new ellipsis button in dashboard edit mode broke the test

JesseSDevaney committed:
adjust test for new fixed width dashboards

JesseSDevaney committed:
adjust test for new fixed width dashboards

JesseSDevaney committed:
adjust test for new fixed width dashboards

JesseSDevaney committed:
adjust test for new fixed width dashboards

JesseSDevaney committed:
Fix double overlay for LineAreaBarChart dash-cards

- No longer has an overlay that persists from editing mode to  viewing mode
- Adjust click-behavior tests to account for new grid-width. Old test relied on a dashcard being taller than they will now default to being. i.e. Move the chosen point/row to down an index.

JesseSDevaney committed:
Merge FixedWidthContainer with dashboard grid div

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

alxnddr committed:
supress brush events while editing timeseries line charts on dashboard

JesseSDevaney committed:
fix failing test

- test failed because I did not update this assertion based on previous changes

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

JesseSDevaney committed:
fix broken link input in editing mode

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

JesseSDevaney committed:
fix pointer events issue on text cards in edit mode

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width

JesseSDevaney committed:
Merge branch 'master' into feature-36358/default-dashboards-to-a-fixed-width
